File: components/src/main/java/com/omvp/components/BottomBarView.java
package com.omvp.components;

import android.content.Context;
import android.os.Build;
import android.support.v4.content.ContextCompat;
import android.support.v4.view.ViewPager;
import android.util.AttributeSet;
import android.util.SparseArray;
import android.view.ViewGroup;
import android.view.ViewOutlineProvider;
import android.widget.LinearLayout;

public class BottomBarView extends BaseComponentView implements BottomBarItemView.OnCheckedChangeListener {

    private LinearLayout mRootView;

    private int[] iconsArray;

    private ViewPager mViewPager;

    private SparseArray<BottomBarItemView> mItemsArray;

    public BottomBarView(Context context) {
        super(context);
    }

    public BottomBarView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public BottomBarView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    protected void loadAttributes(Context context, AttributeSet attrs) {

    }

    @Override
    protected void bindViews() {
        mRootView = findViewById(R.id.root);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            setElevation(getResources().getDimensionPixelSize(R.dimen.elevation));
            setOutlineProvider(ViewOutlineProvider.BOUNDS);
        }
    }

    @Override
    protected void loadData() {
        mItemsArray = new SparseArray<>();
    }

    @Override
    protected int getLayoutId() {
        return R.layout.bottom_bar_view;
    }

    public void setViewPager(ViewPager viewPager, int[] iconsArray) {
        mViewPager = viewPager;
        this.iconsArray = iconsArray;
        addItems();
    }

    private void addItems() {
        LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(
                0,
                ViewGroup.LayoutParams.MATCH_PARENT
        );
        params.weight = 1;

        int numItems = mViewPager.getAdapter().getCount();
        for (int i = 0; i < numItems; i++) {
            BottomBarItemView itemView = new BottomBarItemView(getContext());
/*
                    .icon(getIcon(i))
                    .animateCounter(true)
                    .params(params)
                    .tag(i)
                    .checked(i == 0)
                    .checkedChangeListener(this)
                    .counterColor(ContextCompat.getColor(getContext(), android.R.color.white))
                    .build();
*/
            itemView.setIcon(getIcon(i));
            itemView.setLayoutParams(params);
            itemView.setAnimateCounter(true);
            itemView.setTag(i);
            itemView.setChecked(i == 0);
            itemView.setOnCheckedChangeListener(this);
            itemView.setCounterColor(ContextCompat.getColor(getContext(), android.R.color.white));
            itemView.setCounterBackground(R.drawable.bottom_bar_dot_bg);
            mRootView.addView(itemView);

            mItemsArray.put(i, itemView);
        }
    }

    private int getIcon(int position) {
        if (iconsArray == null || position >= iconsArray.length) {
            return R.drawable.bottom_navigation_item_home;
        }

        return iconsArray[position];
    }

    @Override
    public void onCheckedChanged(BottomBarItemView itemView, boolean isChecked) {
        if (isChecked) {
            for (int i = 0; i < mItemsArray.size(); i++) {
                if ((Integer) itemView.getTag() != i) {
                    BottomBarItemView item = mItemsArray.valueAt(i);
                    item.setChecked(false);
                }
            }

            mViewPager.setCurrentItem((Integer) itemView.getTag(), false);
        }
    }

    public void incrementCounterAt(int position) {
        mItemsArray.valueAt(position).setCounter(1);
    }
}

File: components/src/main/java/com/omvp/components/BottomBarView.kt
package com.omvp.components

import android.content.Context
import android.os.Build
import android.util.AttributeSet
import android.util.SparseArray
import android.view.ViewGroup
import android.view.ViewOutlineProvider
import android.widget.LinearLayout
import androidx.core.content.ContextCompat
import androidx.viewpager.widget.ViewPager

class BottomBarView : BaseComponentView, BottomBarItemView.OnCheckedChangeListener {

    private lateinit var mRootView: LinearLayout

    private var mIconsArray: IntArray? = null

    private var mViewPager: ViewPager? = null

    private var mItemsArray: SparseArray<BottomBarItemView> = SparseArray()

    override val layoutId: Int
        get() = R.layout.bottom_bar_view

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    override fun loadAttributes(context: Context, attrs: AttributeSet?) {

    }

    override fun bindViews() {
        mRootView = findViewById(R.id.root)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            elevation = resources.getDimensionPixelSize(R.dimen.elevation).toFloat()
            outlineProvider = ViewOutlineProvider.BOUNDS
        }
    }

    override fun loadData() {

    }

    fun setViewPager(viewPager: ViewPager, iconsArray: IntArray) {
        mViewPager = viewPager
        mIconsArray = iconsArray
        addItems()
    }

    private fun addItems() {
        val params = LinearLayout.LayoutParams(
                0,
                ViewGroup.LayoutParams.MATCH_PARENT
        )
        params.weight = 1f

        val numItems = mViewPager?.adapter?.count ?: 0

        for (i in 0 until numItems) {
            val itemView = BottomBarItemView(context).also {
                it.setIcon(getIcon(i))
                it.layoutParams = params
                it.setAnimateCounter(true)
                it.tag = i
                it.isChecked = i == 0
                it.setOnCheckedChangeListener(this)
                it.setCounterColor(ContextCompat.getColor(context, android.R.color.white))
                it.setCounterBackground(R.drawable.bottom_bar_dot_bg)
            }

            mRootView.addView(itemView)
            mItemsArray.put(i, itemView)
        }
    }

    private fun getIcon(position: Int): Int {
        return if (mIconsArray == null || position >= mIconsArray!!.size) {
            R.drawable.bottom_navigation_item_home
        } else mIconsArray!![position]

    }

    override fun onCheckedChanged(itemView: BottomBarItemView, isChecked: Boolean) {
        if (isChecked) {
            for (i in 0 until mItemsArray.size()) {
                if (itemView.tag as Int != i) {
                    val item = mItemsArray.valueAt(i)
                    item.isChecked = false
                }
            }

            mViewPager?.setCurrentItem(itemView.tag as Int, false)
        }
    }

    fun incrementCounterAt(position: Int) {
        mItemsArray.valueAt(position).setCounter(1)
    }
}
